### PR TITLE
Add mergeProps argument

### LINKED ISSUE #19: Reintroduce `mergeProps` argument to `connect`
This is to allow connected components to optionally decide how props should be merged.

E.g. use previous value until a live (non-null) value is received from Firebase:

```js
const assignCustomizer = (objectValue, sourceValue) => (
  _.isNil(objectValue) ? sourceValue : objectValue
)

const customMergeProps = (actionProps, subscriptionProps, ownProps) => (
  _.assignWith({}, ownProps, actionProps, subscriptionProps, assignCustomizer)
)
```

### PR DESCRIPTION
**Pull request by simenbrekken**
Closes #19

### CODE REVIEW

**Comment by einarlove** [comment]
Shouldn't actionProps be the last argument seen it will be the least used?

**Comment by simenbrekken** [comment]
Won't all of them almost always be used? Otherwise your function would be discarding arguments.

**Comment by einarlove** [comment]
```
const customMerge = (ownProps, ...rest) => ({
  ...rest,
  prop: ownProps.prop || rest.prop,
})

**Comment by simenbrekken** [comment]
I reordered the arguments in the latest commit, let me know what you think.

**Comment by einarlove** [comment]
Can you give an example where it is necessary to have `subscriptionProps` and `actionProps` as seperate arguments?

On Tue, Jan 24, 2017 at 7:38 AM, Simen Brekken <notifications@github.com> wrote:
I reordered the arguments in the latest commit, let me know what you think.

**Comment by simenbrekken** [comment]
We could merge them before passing them into `mergeProps` as they will never collide, but an efficient cascading implementation would still be:

```js
const assignCustomizer = (objectValue, sourceValue) => (
  _.isNil(objectValue) ? sourceValue : objectValue
)

const customMergeProps = (actionProps, subscriptionProps, ownProps) => {
  const cascadedProps = _.assignWith({}, ownProps, subscriptionProps, assignCustomizer)

  return {
    ...actionProps,
    ...cascadedProps
  }
}
```

File: src/connect.js [hunk 1]
```diff
@@ -129,10 +129,10 @@ export default (mapFirebaseToProps = defaultMapFirebaseToProps) => {
       }
 
       render() {
+        const subscriptionProps = this.state.subscriptionsState
         const firebaseProps = mapFirebase(this.props, this.ref, this.firebaseApp)
         const actionProps = pickBy(firebaseProps, isFunction)
-        const subscriptionProps = this.state.subscriptionsState
-        const props = mergeProps(actionProps, subscriptionProps, this.props)
+        const props = mergeProps(this.props, subscriptionProps, actionProps)
```
Comment (einarlove) [inline src/connect.js]: Why can't we do
```js
const props = mergeProps(this.props, { ...actionProps, ...subscriptionProps })
```
instead? Why should the developer get three arguments to `mapProps`?

Comment (simenbrekken) [inline src/connect.js]: If you'd want to add defaults to `subscriptionProps` as in the example above you would have to filter out `actionProps` by checking if they are functions or values.

Comment (einarlove) [inline src/connect.js]: Could you give an example?

Comment (simenbrekken) [inline src/connect.js]: https://github.com/unfold/react-firebase/pull/20#issuecomment-274747301

**Comment by simenbrekken** [comment]
Once again, Circle refuses to run tests so I'm merging this manually.

**Pull request merged**